src/generator/qbe.rs

```diff
@@ -0,0 +1,596 @@
+        }
+    }
+}
+
+/// Function block with a label
+#[derive(Debug)]
+struct QbeBlock {
+    /// Label before the block
+    label: String,
+
+    /// A list of instructions in the block
+    instructions: Vec<QbeExpr>,
+}
+
+impl QbeBlock {
+    /// Adds a new instruction to the block
+    fn add_instr(&mut self, instr: QbeInstr) {
+        self.instructions.push(QbeExpr::Volatile(instr));
+    }
+
+    /// Adds a new instruction assigned to a temporary
+    fn assign_instr(&mut self, temp: QbeTemporary, ty: QbeType, instr: QbeInstr) {
+        self.instructions.push(QbeExpr::Assign(temp, ty, instr));
+    }
+}
+
+impl fmt::Display for QbeBlock {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        writeln!(f, "@{}", self.label)?;
+
+        write!(
+            f,
+            "{}",
+            self.instructions
+                .iter()
+                .map(|instr| format!("\t{}", instr))
+                .collect::<Vec<String>>()
+                .join("\n")
+        )
+    }
+}
+
+/// QBE function
+#[derive(Debug)]
+struct QbeFunction {
+    /// Should the function be available to outside users
+    exported: bool,
+
+    /// Function name
+    name: String,
+
+    /// Function arguments
+    arguments: Vec<(QbeType, QbeTemporary)>,
+
+    /// Return type
+    return_ty: Option<QbeType>,
+
+    /// Labelled blocks
+    blocks: Vec<QbeBlock>,
+}
+
+impl QbeFunction {
+    /// Adds a new empty block with a specified label
+    fn add_block(&mut self, label: String) {
+        self.blocks.push(QbeBlock {
+            label,
+            instructions: Vec::new(),
+        });
+    }
+
+    /// Adds a new instruction to the last block
+    fn add_instr(&mut self, instr: QbeInstr) {
+        self.blocks
+            .last_mut()
+            .expect("Last block must be present")
+            .add_instr(instr);
+    }
+
+    /// Adds a new instruction assigned to a temporary
+    fn assign_instr(&mut self, temp: QbeTemporary, ty: QbeType, instr: QbeInstr) {
+        self.blocks
+            .last_mut()
+            .expect("Last block must be present")
+            .assign_instr(temp, ty, instr);
+    }
+}
+
+impl fmt::Display for QbeFunction {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.exported {
+            write!(f, "export ")?;
+        }
+        write!(f, "function")?;
+        if let Some(ty) = &self.return_ty {
+            write!(f, " {}", ty)?;
+        }
+
+        writeln!(
+            f,
+            " ${name}({args}) {{",
+            name = self.name,
+            args = self
+                .arguments
+                .iter()
+                .map(|arg| format!("{} {}", arg.0, arg.1))
+                .collect::<Vec<String>>()
+                .join(", "),
+        )?;
+
+        for blk in self.blocks.iter() {
+            writeln!(f, "{}", blk)?;
+        }
+
+        write!(f, "}}")
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Just for the sake of readability, we should probably extract the tests into their own file in `tests/`.

Reply:
I don't think this will work because `Qbe*` data structures are private